src/main/java/gregtech/api/capability/impl/CleanroomLogic.java

```diff
@@ -0,0 +1,228 @@
+package gregtech.api.capability.impl;
+
+import gregtech.api.GTValues;
+import gregtech.api.capability.GregtechDataCodes;
+import gregtech.api.metatileentity.MetaTileEntity;
+import gregtech.api.metatileentity.multiblock.ICleanroomProvider;
+import gregtech.api.metatileentity.multiblock.IMaintenance;
+import gregtech.common.ConfigHolder;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.network.PacketBuffer;
+
+import javax.annotation.Nonnull;
+
+public class CleanroomLogic {
+
+    public static final int MAX_PROGRESS = 100; // five seconds
+    public static final int BASE_CLEAN_AMOUNT = 5;
+
+    private int progressTime = 0;
+
+    private final int minEnergyTier;
+
+    private final MetaTileEntity metaTileEntity;
+    private final boolean hasMaintenance;
+
+    private boolean isActive;
+    private boolean isWorkingEnabled = true;
+    private boolean wasActiveAndNeedsUpdate;
+
+    private boolean hasNotEnoughEnergy;
+
+    public CleanroomLogic(MetaTileEntity metaTileEntity, int minEnergyTier) {
+        this.metaTileEntity = metaTileEntity;
+        this.minEnergyTier = minEnergyTier;
+        this.hasMaintenance = ConfigHolder.machines.enableMaintenance && ((IMaintenance) metaTileEntity).hasMaintenanceMechanics();
+    }
+
+    /**
+     * Performs the actual cleaning
+     * Call this method every tick in update
+     */
+    public void updateLogic() {
+        if (metaTileEntity.getWorld().isRemote) return;
+
+        if (hasMaintenance && ((IMaintenance) metaTileEntity).getNumMaintenanceProblems() > 5) {
+            return;
+        }
+
+        // cleanrooms which cannot work do nothing
+        if (!this.isWorkingEnabled) return;
+
+        // drain the energy
+        consumeEnergy(false);
+
+        if (!this.isActive) setActive(true);
+
+        // increase progress
+        progressTime++;
+        if (progressTime % getMaxProgress() != 0) return;
+        progressTime = 0;
+
+        int amountToClean;
+        if (hasEnoughPower()) amountToClean = BASE_CLEAN_AMOUNT * (getTierDifference() + 1);
+        else amountToClean = -BASE_CLEAN_AMOUNT;
+
+        // each maintenance problem lowers gain by 1 and increases loss by 1
+        if (hasMaintenance) amountToClean -= ((IMaintenance) metaTileEntity).getNumMaintenanceProblems();
+        ((ICleanroomProvider) metaTileEntity).adjustCleanAmount(amountToClean);
+    }
+
+    protected boolean consumeEnergy(boolean simulate) {
+        return ((ICleanroomProvider) metaTileEntity).drainEnergy(simulate);
+    }
+
+    /**
+     * @return true if the cleanroom is able to drain energy, else false
+     */
+    protected boolean hasEnoughPower() {
+        if (!consumeEnergy(true)) {
```

Comment:
If my previous suggestion is taken, this check would not be needed. Also, it does not quite make sense. Right now the structure always drains energy, even if it may not have enough energy in the first place, but then does a simulate check here?